backend/pokerogue/src/main/java/com/pokerogue/helper/ability/service/AbilityService.java

```diff
@@ -20,43 +20,39 @@
 public class AbilityService {
 
     private final S3Service s3Service;
-    private final InMemoryAbilityRepository inMemoryAbilityRepository;
+    private final AbilityRepository abilityRepository;
+    private final PokemonRepository pokemonRepository;
 
     public List<AbilityResponse> findAbilities() {
-        return inMemoryAbilityRepository.findAll().stream()
+        return abilityRepository.findAll().stream()
+                .filter(Ability::isNotNone)
                 .map(AbilityResponse::from)
                 .toList();
     }
 
     public AbilityDetailResponse findAbilityDetails(String id) {
-        Ability ability = inMemoryAbilityRepository.findById(id)
+        Ability ability = abilityRepository.findById(id)
                 .orElseThrow(() -> new GlobalCustomException(ErrorMessage.POKEMON_ABILITY_NOT_FOUND));
-        List<AbilityPokemonResponse> abilityPokemonResponses = ability.getInMemoryPokemons().stream()
-                .map(abilityPokemon -> new AbilityPokemonResponse(
-                        abilityPokemon.id(),
-                        Long.parseLong(abilityPokemon.speciesId()),
-                        abilityPokemon.koName(),
-                        s3Service.getPokemonImageFromS3(abilityPokemon.id()),
-                        getAbilityTypeResponses(abilityPokemon.firstType(), abilityPokemon.secondType())
+        List<String> abilityPokemonIds = ability.getPokemonIds();
+        List<AbilityPokemonResponse> abilityPokemonResponses = pokemonRepository.findAllById(abilityPokemonIds).stream()
+                .map(pokemon -> new AbilityPokemonResponse(
+                        pokemon.getId(),
+                        (long) pokemon.getPokedexNumber(),
+                        pokemon.getKoName(),
+                        s3Service.getPokemonImageFromS3(pokemon.getImageId()),
+                        getAbilityTypeResponses(pokemon.getTypes())
                 ))
                 .toList();
+        if (abilityPokemonIds.size() == abilityPokemonResponses.size()) {
```

Comment:
스트림에서 사이즈가 변할만한 메서드가 없는데요. 
이 조건식이 필요할까요 🐧

Reply:
`MongoRepository`의 `findAllById` 메소드는 `List<String>`으로 전달한 `id` 목록 중 존재하지 않는 `id`가 있으면 예외를 처리하는 게 아닌 그 값은 건너뛰고 반환합니다! 우리 서비스에서는 건너뛴 상태로 반환해선 안된다고 판단해 해당 로직을 추가했습니다!

Reply:
```
Returns all instances of the type T with the given IDs.
If some or all ids are not found, no entities are returned for these IDs.
Note that the order of elements in the result is not guaranteed.
Params:
ids – must not be null nor contain any null values.
Returns:
guaranteed to be not null. The size can be equal or less than the number of given ids.
Throws:
IllegalArgumentException – in case the given ids or one of its items is null.
```

`findAllById`
리스트의 원소 중에서 `null`이 있으면 예외가 터지고, 찾을 수 없으면 아무것도 반환되지 않는군요.
검증이 한단계 더 추가되는데,  `findById`를 사용하지 않고 `findAllById`를 사용한 이유가 있을까요?

Reply:
반환되는 사이즈를 알 수 없으니까 사이즈 검증을 하고 있는 것 아닌가요 ...??

포켓몬 아이디가 3개인데 2개를 반환하면 아이디 목록이 잘못된 거니까요 🤔

Findbyid 해서 optional이 empty인지 확인하는 것보다 in 쿼리를 날려서 사이즈 검증을 하는 것이 검증을 한 번 더 하는 걸까요?

오히려 모든 id에 대해서 optional이 empty인지 확인하는 게 검증을 더 많이 하는 것 아닌가요??

Reply:
`Null` 반환의 경우 사실 전부 `Get` 해주는 과정이 있기 때문에 어느 방식이든 모두 `NPE`가 발생할 것 같다는 생각이 들어요.

각각의 장단점이 있다고 생각되네요 🤔

`findById` 는 단건 조회를 함으로써 만약 데이터에 문제가 생긴다면 정확히 어떤 데이터가 문제를 발생시키는 지 쉽게 알 수 있을 것 같아요. 또 그냥 바로바로 `Dto`로 만들어도 되겠네요. ☺️
다만 하나의 API에서 가져오려는 데이터만큼 데이터베이스를 여러번 호출해야한다는 단점이 있을 것 같아요. 

`findAllById`는 단 한번의 데이터베이스 호출만 한다는 장점이 생기겠네요. 
다만, 과정에서 문제가 생긴다면 정확히 어떤 데이터에서 문제가 생긴것인지 알기가 어려울 것 같다는 생각이 듭니다. ☺️

사실 어떻게 보면 저희의 데이터는 저희끼리 형식을 정하기도 했고, 해당 데이터에서는 `Null` 이 들어가는 경우가 드물것 같아서
후자에 좀 더 마음이 가긴 합니다.

## 제안

아니면 이건 제안인데 어쩌피 저희 그 쿼리 다듬고 서비스 정리해서 개선 해야하지 않나요??
일단 단건 조회를 하고 10만건이나 100만건 데이터 집어넣고 본격적으로 비교해보는 것도 나쁘지 않을수도???

얼마 차이 안나면 사실 단건 조회해도 괜찮을 거고 너무 많은 시간을 잡아먹고 있거나 데이터 연결을 긴 시간동안 들고 있다고 판단되면 여러건 조회로 바꿔도 괜찮을 것 같아요! 😯

Reply:
종이가 둘다 상관 없다네요! 
그러면 그대로 가죠???